{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# intro \n",
    "\n",
    "*slack message Paolo*:\n",
    "\n",
    "OK in order to be helpful I needed to take a look myself.\n",
    "\n",
    "I suggest structuring the work in these parts:\n",
    "\n",
    "1. Understand the generative process (no real data, just generation based on your mental model of thigs). I did this in the script attached \n",
    "    (my plots look different from yours)\n",
    "    \n",
    "2. Make sure your data follow the hypothesized model. specifically, in the considered model you assume that the data come from beta_binomial, where n is linear in age, while mean_beta and var_beta do not depend on age. Is this true? You can check it by (i) plot coverage vs age (is this linear?); (ii) fit beta binomials on young and old people groups separately and see how the MLEs of mean_beta and var_beta correlate across genes.\n",
    "    \n",
    "3. If this looks good -> your model is reasonable -> you can assume things are fine. If you see deviations you need to rethinkg about the generative processs---e.g., may var_beta also depend on age?\n",
    "\n",
    "\n",
    "*My thoughts on this:*\n",
    "\n",
    "1. Absolutely, and first and foremost thanks a lot for the clarity and advice. Indeed, I could reproduce data and plots which looked like yours. The difference is, as you've already pointed out below, if one makes the variance of the Beta distribution dependent on coverage or not. In the code you've sent, the were not linked. Below they are, as well as in all the scenarios I showed in previous meetings. \n",
    "\n",
    "2. I went back to the original GTEx heteroplasmy data again and found this was not super clear. I have some slides to show this if you're interested, but in brief: (i) there seems to be some modest relationship between coverage and age overall. And it's in the ball park of being linear. (ii) for the positions the linear model seemed to pick up on when testing for relationship between heteroplasmy and donor age, the relationship looked linear as well (analytical p-values; study-wide bonferroni). (iii) for the same postions, there definitely seems to be an inverse relationship between the variance in heteroplasmy estimates and coverage. *In conclusion, this indicates to me, that we should link coverage with variance in the Beta distribution, which is what I did below.* Age vs. coverage may not be linear relationship in the real data. But as long as it induces an inverse relationship between variance in heteroplasmy and coverage which is also correlated with age, this seems fair to me.\n",
    "\n",
    "3. I didn't manage to fit the models yet. This is ongoing and I'll update you on that."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# code\n",
    "## init"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select python environment `gene_tools`\n",
    "# source deps\n",
    "import numpy as np\n",
    "import pdb\n",
    "import pylab as plt\n",
    "import pandas as pd\n",
    "import os\n",
    "import random\n",
    "import session_info\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "from scipy.stats import truncnorm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## define functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2024_01_31: let's turn interactive code into a function\n",
    "def sample_cohort_specs(n_donors = 500, mean_age = 50, sd_age = 10, \n",
    "                        min_age_bound = 20, max_age_bound = 70, \n",
    "                        min_coverage = 10,  r_sqr = 0.1, \n",
    "                        slope = 0.1, mean_coverage=1000):\n",
    "    \"\"\"\n",
    "    Generate synthetic cohort similar to GTEx data. The main point is to simulate\n",
    "    coverage values which are linearly dependent on donor age. The parameter \n",
    "    values are chosen such that the properties of the simulated cohort match\n",
    "    those of GTEx tissues best. These simulations are best thought of as \"per \n",
    "    tissue, per position\". \n",
    "\n",
    "    - donor ages are simulated using a truncated normal. \n",
    "    - coverage values are simulated using a linear model and Gaussian noise.\n",
    "    \n",
    "\n",
    "    n_donors:               number of donors to simulate \n",
    "    mean_age:               mean of the resulting age distribution\n",
    "    sd_age:                 standard deviation of the resulting age distribution\n",
    "    min_age_bound:          minimum of donor age distribution to simulate\n",
    "    max_age_bound:          minimum of donor age distribution to simulate\n",
    "    slope:                  slope of age coverage linear relationship.  \n",
    "    r_sqr:                  R^2 value specified for age ~ coverage relationship.\n",
    "    mean_coverage:          mean of coverage values to aim for when simulating.\n",
    "    min_coverage:           minimum coverage value found in real data.\n",
    "\n",
    "    \n",
    "    returns:   \"synthetic cohort\"  for tissue X and one position. Object is PD\n",
    "               containing columns age, mean_age, coverage, mean_coverage, coverage.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # simulate cohort ----------------------------------------------------------\n",
    "    # init empty data frame\n",
    "    sim_df = pd.DataFrame(np.zeros(n_donors),columns=[\"donor_id\"])\n",
    "    # generate donor_id \n",
    "    sim_df['donor_id'] = 'donor_' + (sim_df.index + 1).astype(str)\n",
    "\n",
    "\n",
    "    # simulate donor ages ------------------------------------------------------\n",
    "    # calculate the truncated normal bounds\n",
    "    a = (min_age_bound - mean_age) / sd_age\n",
    "    b = (max_age_bound - mean_age) / sd_age\n",
    "    # generate ages from a truncated normal distribution\n",
    "    ages = truncnorm.rvs(a, b, loc = mean_age, scale = sd_age, size = n_donors)\n",
    "    # round ages to integers\n",
    "    simulated_age = np.round(ages).astype(int)\n",
    "\n",
    "\n",
    "    # simulate coverage in linear dependence of donor age ----------------------\n",
    "    # the key question here is how to relate mean_coverage to our linear model\n",
    "    # slope coverage is defined as the [%] of change per year of life\n",
    "    # delete:  slope_coverage = mean_coverage * slope_gradient \n",
    "    # slope is fixed and intercept needs to be computed\n",
    "    #     beta = simulated_age * slope_coverage\n",
    "    intercept = mean_coverage - np.mean(simulated_age) * slope\n",
    "    var_coverage = (((1 - r_sqr)/r_sqr) * slope**2) * np.var(simulated_age)\n",
    "    sd_coverage = np.sqrt(var_coverage)\n",
    "    simulated_coverage = intercept + simulated_age * slope + np.random.normal(loc = 0, scale = sd_coverage, size = len(sim_df)) \n",
    "    \n",
    "    \n",
    "    # for coverage < min_coverage repeat simulation  \n",
    "    simulated_coverage = np.maximum(simulated_coverage, min_coverage)  \n",
    "    # continue sampling if there are still values smaller than min_coverage\n",
    "    while np.any(simulated_coverage <= min_coverage):\n",
    "        # resample only the values that are smaller than min_coverage\n",
    "        resample_mask = simulated_coverage <= min_coverage\n",
    "        resample_size = np.sum(resample_mask)\n",
    "        # generate new values for the resampled subset\n",
    "        new_values = intercept + simulated_age[resample_mask] * slope + np.random.normal(loc = 0, scale = sd_coverage, size = resample_size)\n",
    "        # update the simulated_coverage array with the new values\n",
    "        simulated_coverage[resample_mask] = np.maximum(new_values, min_coverage)\n",
    "\n",
    "    # add return values to synthetic cohorts -----------------------------------\n",
    "    sim_df[\"age\"] = simulated_age\n",
    "    sim_df[\"mean_age\"] = np.mean(sim_df[\"age\"])\n",
    "    sim_df[\"coverage_non_rounded\"] = simulated_coverage\n",
    "    sim_df[\"mean_coverage\"] = np.mean(sim_df[\"coverage_non_rounded\"])\n",
    "    sim_df[\"coverage\"] = np.round(simulated_coverage).astype(int)\n",
    "\n",
    "\n",
    "    # return result ------------------------------------------------------------\n",
    "    return sim_df    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2024_01_30: cleanup\n",
    "def sample_beta_binomial_fixed_variance(sim_df, mean_beta=0.5, var_beta=0.1):\n",
    "    \"\"\"\n",
    "    uses simulated coverage values from linear age ~ coverage relationship and \n",
    "    draw synthetic heteroplasmy values using a BetaBinomial only dependent on \n",
    "    coverage.\n",
    "\n",
    "    mean_beta:  mean to be used in the Beta function.\n",
    "    var_beta:   variance to be used in the Beta function.\n",
    "    \n",
    "    returns:    synthetic heteroplasmy counts/levels attached as cols to input \n",
    "                synthetic cohort PD (generated 1 step above).\n",
    "\n",
    "    \"\"\"\n",
    "    # use constant coverage\n",
    "    sim_df[\"var_beta\"] = var_beta\n",
    "\n",
    "    # set mean to expected mean_het\n",
    "    sim_df[\"mean_het\"] = mean_beta\n",
    "\n",
    "    # calculate alpha and beta from mean and variance\n",
    "    sim_df[\"common_factor\"] = sim_df[\"mean_het\"] * (1 - sim_df[\"mean_het\"]) / sim_df[\"var_beta\"] - 1\n",
    "    sim_df[\"alpha\"] = sim_df[\"mean_het\"] * sim_df[\"common_factor\"]\n",
    "    sim_df[\"beta\"] = (1 - sim_df[\"mean_het\"]) * sim_df[\"common_factor\"]\n",
    "\n",
    "    # 2024_01_01: adding pseudocount to ensure alpha and beta are positive\n",
    "    sim_df[\"alpha\"] = np.maximum(sim_df[\"alpha\"], 1e-6)  # Set a minimum value, adjust as needed\n",
    "    sim_df[\"beta\"] = np.maximum(sim_df[\"beta\"], 1e-6)\n",
    "\n",
    "    # sample from p_beta from Beta for using it in Bernoulli draw below\n",
    "    sim_df[\"p_beta\"] = np.random.beta(sim_df[\"alpha\"],sim_df[\"beta\"])\n",
    "\n",
    "    # sample from the Binomial distribution using the sampled Beta value\n",
    "    sim_df[\"het_counts\"] = np.random.binomial(sim_df[\"coverage\"], sim_df[\"p_beta\"])\n",
    "    sim_df[\"het_level\"] = sim_df[\"het_counts\"]/sim_df[\"coverage\"]\n",
    "\n",
    "    return sim_df       "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## apply functions -- manual for testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set params manually for debugging before running automated pipeline above\n",
    "## age ~ coverage params\n",
    "n_donors = 500\n",
    "mean_age = 50\n",
    "sd_age = 15\n",
    "min_age_bound = 20\n",
    "max_age_bound = 70\n",
    "min_coverage = 10\n",
    "\n",
    "# alternate to find spot from graphs which combinations are plausible\n",
    "r_sqr = 0.1\n",
    "slope = 350\n",
    "mean_coverage = 20000\n",
    "## heteroplasmy params\n",
    "mean_beta = 0.5\n",
    "var_beta = 0.01\n",
    "\n",
    "## set N positions to simulate\n",
    "n_pos = 100\n",
    "\n",
    "# init res list\n",
    "sim_list = []\n",
    "# loop through the number of positions you want to simulate\n",
    "for pos in range(1, n_pos + 1):\n",
    "    # simulate coverage\n",
    "    tmp_df = sample_cohort_specs(n_donors=n_donors,\n",
    "                                 mean_age=mean_age,\n",
    "                                 sd_age=sd_age,\n",
    "                                 min_age_bound=min_age_bound,\n",
    "                                 max_age_bound=max_age_bound,\n",
    "                                 min_coverage=min_coverage,\n",
    "                                 r_sqr=r_sqr,\n",
    "                                 slope=slope, \n",
    "                                 mean_coverage=mean_coverage) \n",
    "    # simulate heteroplasmy inversely correlated with coverage using beta binomial\n",
    "    tmp_df = sample_beta_binomial_fixed_variance(sim_df=tmp_df, \n",
    "                                                 mean_beta=mean_beta, \n",
    "                                                 var_beta=var_beta)\n",
    "    # add Pos column \n",
    "    tmp_df[\"Pos\"] = pos\n",
    "    # append the result to the list\n",
    "    sim_list.append(tmp_df)\n",
    "\n",
    "# concatenate the results into a single DataFrame\n",
    "sim_df = pd.concat(sim_list, ignore_index=True)\n",
    "\n",
    "# show contents sim_df \n",
    "#sim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot mean_het level\n",
    "mean_het_level = np.mean(sim_df[\"het_level\"])\n",
    "\n",
    "# make plots for sanity checking \n",
    "plt.figure(1, figsize=(18, 5))\n",
    "\n",
    "# # 1\n",
    "plt.subplot(131)\n",
    "plt.plot(sim_df[\"age\"], sim_df[\"coverage\"], '.k')\n",
    "plt.title('simulations: age vs. coverage \\n', fontsize = 16, fontweight = \"bold\")\n",
    "plt.xlabel('age [y]', fontsize = 14, fontweight = \"bold\")\n",
    "plt.ylabel('coverage [reads]', fontsize = 14, fontweight = \"bold\")\n",
    "\n",
    "# # 2\n",
    "plt.subplot(132)\n",
    "plt.plot(sim_df[\"age\"], sim_df[\"het_level\"], '.k')\n",
    "plt.axhline(y=mean_het_level, color='r', linestyle='-', label=f'Mean Het Level: {mean_het_level:.2f}')\n",
    "plt.title(f'simulations: coverage vs. heteroplasmy \\n(Mean het_level simulated: {mean_het_level:.5f})', fontsize = 16, fontweight = \"bold\")\n",
    "#plt.title('simulations: age vs. heteroplasmy', fontsize = 16, fontweight = \"bold\")\n",
    "plt.ylabel('heteroplasmy [%]', fontsize = 14, fontweight = \"bold\")\n",
    "plt.xlabel('age [y]', fontsize = 14, fontweight = \"bold\")\n",
    "plt.ylim(0, 1)\n",
    "\n",
    "# # 3\n",
    "plt.subplot(133)\n",
    "plt.plot(sim_df[\"coverage\"], sim_df[\"het_level\"], '.k')\n",
    "plt.axhline(y=mean_het_level, color='r', linestyle='-', label=f'Mean Het Level: {mean_het_level:.2f}')\n",
    "plt.ylabel('heteteroplasmy [%]', fontsize = 14, fontweight = \"bold\")\n",
    "plt.xlabel('coverage [reads]', fontsize = 14, fontweight = \"bold\")\n",
    "#plt.title('simulations: coverage vs. heteroplasmy', fontsize = 16, fontweight = \"bold\")\n",
    "plt.title(f'simulations: coverage vs. heteroplasmy \\n(Mean heat_level simulated: {mean_het_level:.5f})', fontsize = 16, fontweight = \"bold\")\n",
    "plt.ylim(0, 1)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_df[\"het_level\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assuming sim_df is a pandas DataFrame\n",
    "correlation_matrix = np.corrcoef(sim_df[\"coverage_non_rounded\"], sim_df[\"age\"])\n",
    "\n",
    "# The correlation coefficient is at position (0, 1) or (1, 0) in the matrix\n",
    "correlation_coefficient = correlation_matrix[0, 1]\n",
    "\n",
    "print(\"Correlation Coefficient:\", correlation_coefficient)\n",
    "print(\"sqrt: Correlation Coefficient:\", np.sqrt(correlation_coefficient))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## apply functions -- manual and motivated by observed but extreme data values\n",
    "\n",
    "### read in observed but extreme values for simulating 8 cohorts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params_df = pd.read_excel(\"/Users/simon.wengert/git/mtDNA_variants/metadata/params/2024_01_30_het_sim_params.xlsx\")\n",
    "params_df = params_df[params_df['run'] == True]\n",
    "params_df[\"n_donors\"] = params_df[\"n_donors\"].astype(int)\n",
    "# params_df.insert(4, 'simulation', range(1, len(params_df) + 1))\n",
    "\n",
    "params_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### for each of these cohorts, assess resulting simulated heterplasmy levels when changing `mean_beta` and `var_beta`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create an empty list to store simulated data frames\n",
    "simulated_dfs = []\n",
    "n_pos = 100\n",
    "# Create a PDF to save all the plots\n",
    "params_df.iloc[0]\n",
    "# Retrieve params\n",
    "row_id = row[\"row_id\"]\n",
    "change_param = row[\"change_param\"]\n",
    "step_param = row[\"step_param\"]\n",
    "n_donors = row['n_donors']\n",
    "mean_age = row[\"mean_age\"]\n",
    "sd_age = row[\"sd_age\"]\n",
    "min_age_bound = row[\"min_age_bound\"]\n",
    "max_age_bound = row[\"max_age_bound\"]\n",
    "min_coverage = row[\"min_coverage\"]\n",
    "slope = row[\"slope\"]\n",
    "r_sqr = row[\"r_sqr\"]\n",
    "mean_coverage = row[\"mean_coverage\"]\n",
    "\n",
    "\n",
    "## heteroplasmy params\n",
    "mean_beta = 0.025\n",
    "var_beta = 0.000001\n",
    "\n",
    "## set N positions to simulate\n",
    "n_pos = 100\n",
    "\n",
    "# init res list\n",
    "sim_list = []\n",
    "# loop through the number of positions you want to simulate\n",
    "for pos in range(1, n_pos + 1):\n",
    "    # simulate coverage\n",
    "    tmp_df = sample_cohort_specs(n_donors=n_donors,\n",
    "                                 mean_age=mean_age,\n",
    "                                 sd_age=sd_age,\n",
    "                                 min_age_bound=min_age_bound,\n",
    "                                 max_age_bound=max_age_bound,\n",
    "                                 min_coverage=min_coverage,\n",
    "                                 r_sqr=r_sqr,\n",
    "                                 slope=slope, \n",
    "                                 mean_coverage=mean_coverage) \n",
    "    # simulate heteroplasmy inversely correlated with coverage using beta binomial\n",
    "    tmp_df = sample_beta_binomial_fixed_variance(sim_df=tmp_df, \n",
    "                                                 mean_beta=mean_beta, \n",
    "                                                 var_beta=var_beta)\n",
    "    # add Pos column \n",
    "    tmp_df[\"Pos\"] = pos\n",
    "    # append the result to the list\n",
    "    sim_list.append(tmp_df)\n",
    "\n",
    "# concatenate the results into a single DataFrame\n",
    "sim_df = pd.concat(sim_list, ignore_index=True)\n",
    "\n",
    "\n",
    "sim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 0 based index which means the thing below is XXX\n",
    "row = params_df.iloc[2] \n",
    "row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create an empty list to store simulated data frames\n",
    "simulated_dfs = []\n",
    "n_pos = 100\n",
    "# Create a PDF to save all the plots\n",
    "\n",
    "# Retrieve params\n",
    "row_id = row[\"row_id\"]\n",
    "change_param = row[\"change_param\"]\n",
    "step_param = row[\"step_param\"]\n",
    "n_donors = row['n_donors']\n",
    "mean_age = row[\"mean_age\"]\n",
    "sd_age = row[\"sd_age\"]\n",
    "min_age_bound = row[\"min_age_bound\"]\n",
    "max_age_bound = row[\"max_age_bound\"]\n",
    "min_coverage = row[\"min_coverage\"]\n",
    "slope = row[\"slope\"]\n",
    "r_sqr = row[\"r_sqr\"]\n",
    "mean_coverage = row[\"mean_coverage\"]\n",
    "\n",
    "# Function to calculate mean heteroplasmy for given mean_beta and var_beta\n",
    "def calculate_mean_het(mean_beta, var_beta):\n",
    "    sim_list = []\n",
    "    for pos in range(1, n_pos + 1):\n",
    "        tmp_df = sample_cohort_specs(n_donors=n_donors,\n",
    "                                     mean_age=mean_age,\n",
    "                                     sd_age=sd_age,\n",
    "                                     min_age_bound=min_age_bound,\n",
    "                                     max_age_bound=max_age_bound,\n",
    "                                     min_coverage=min_coverage,\n",
    "                                     r_sqr=r_sqr,\n",
    "                                     slope=slope,\n",
    "                                     mean_coverage=mean_coverage)\n",
    "        tmp_df = sample_beta_binomial_fixed_variance(sim_df=tmp_df,\n",
    "                                                     mean_beta=mean_beta,\n",
    "                                                     var_beta=var_beta)\n",
    "        tmp_df[\"Pos\"] = pos\n",
    "        sim_list.append(tmp_df)\n",
    "\n",
    "    sim_df = pd.concat(sim_list, ignore_index=True)\n",
    "    return sim_df[\"het_level\"].mean()\n",
    "\n",
    "\n",
    "# init grid for comparison:\n",
    "mean_beta_values = [0.0, 0.01, 0.02, 0.05, 0.1, 0.2, 0.3, 0.4,  0.5, 0.75, 0.9]\n",
    "# var_beta_values = [0.000001, 0.00001, 0.0001, 0.001, 0.01, 0.1,1.0,10]\n",
    "var_beta_values = [0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05, 0.1]\n",
    "\n",
    "\n",
    "\n",
    "# Create an empty dictionary to store results\n",
    "result_dict = {}\n",
    "\n",
    "# Iterate over the grid\n",
    "for mean_beta in mean_beta_values:\n",
    "    for var_beta in var_beta_values:\n",
    "        # Calculate het_level\n",
    "        current_het_level = calculate_mean_het(mean_beta, var_beta)\n",
    "        \n",
    "        # Store the result in the dictionary\n",
    "        result_dict[(mean_beta, var_beta)] = current_het_level\n",
    "\n",
    "# Print the results\n",
    "for (mean_beta, var_beta), het_level in result_dict.items():\n",
    "    print(f'Mean Beta: {mean_beta}, Var Beta: {var_beta}, Het Level: {het_level}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make heatmap\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Convert the result_dict to a DataFrame for easier plotting\n",
    "result_df = pd.DataFrame(list(result_dict.items()), columns=['Beta', 'Het_Level'])\n",
    "\n",
    "# Split the 'Beta' column into two separate columns: 'Mean_Beta' and 'Var_Beta'\n",
    "result_df[['Mean_Beta', 'Var_Beta']] = pd.DataFrame(result_df['Beta'].tolist(), index=result_df.index)\n",
    "\n",
    "# Create a pivot table for plotting the heatmap\n",
    "heatmap_data = result_df.pivot_table(index='Var_Beta', columns='Mean_Beta', values='Het_Level')\n",
    "\n",
    "# Plot the heatmap\n",
    "plt.figure(figsize=(10, 8))\n",
    "sns.heatmap(heatmap_data, cmap='viridis', annot=True, fmt=\".4f\", cbar_kws={'label': 'Het Level'})\n",
    "plt.title('Het Level for Mean Beta vs Var Beta')\n",
    "plt.xlabel('Mean Beta')\n",
    "plt.ylabel('Var Beta')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the result_dict to a DataFrame for easier plotting\n",
    "result_df = pd.DataFrame(list(result_dict.items()), columns=['Beta', 'Het_Level'])\n",
    "\n",
    "# Split the 'Beta' column into two separate columns: 'Mean_Beta' and 'Var_Beta'\n",
    "result_df[['Mean_Beta', 'Var_Beta']] = pd.DataFrame(result_df['Beta'].tolist(), index=result_df.index)\n",
    "\n",
    "# Plot a scatter plot with log scale on both axes, larger dots, and legend outside the plot\n",
    "plt.figure(figsize=(12, 8))\n",
    "#scatter_plot = sns.scatterplot(data=result_df, x='Mean_Beta', y='Var_Beta', hue='Het_Level', palette='viridis', s=500)\n",
    "scatter_plot = sns.scatterplot(data=result_df, x='Mean_Beta', y='Var_Beta', hue='Het_Level', palette='viridis', s=700, marker='s')\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')\n",
    "plt.title('Het Level for Mean Beta vs Var Beta (log scale)', fontsize=16, fontweight='bold')\n",
    "plt.xlabel('Mean Beta (log scale)', fontsize=14, fontweight='bold')\n",
    "plt.ylabel('Var Beta (log scale)', fontsize=14, fontweight='bold')\n",
    "plt.legend(title='Het Level', bbox_to_anchor=(1.05, 1), loc='upper left', fontsize='large', title_fontsize='large')\n",
    "\n",
    "# Increase font size and boldness for axis labels\n",
    "scatter_plot.set_xticklabels(scatter_plot.get_xticks(), fontsize=12, fontweight='bold')\n",
    "scatter_plot.set_yticklabels(scatter_plot.get_yticks(), fontsize=12, fontweight='bold')\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "import numpy as np\n",
    "\n",
    "# Function to calculate mean heteroplasmy for given mean_beta and var_beta\n",
    "def calculate_mean_het(mean_beta, var_beta):\n",
    "    sim_list = []\n",
    "    for pos in range(1, n_pos + 1):\n",
    "        tmp_df = sample_cohort_specs(n_donors=n_donors,\n",
    "                                     mean_age=mean_age,\n",
    "                                     sd_age=sd_age,\n",
    "                                     min_age_bound=min_age_bound,\n",
    "                                     max_age_bound=max_age_bound,\n",
    "                                     min_coverage=min_coverage,\n",
    "                                     r_sqr=r_sqr,\n",
    "                                     slope=slope,\n",
    "                                     mean_coverage=mean_coverage)\n",
    "        tmp_df = sample_beta_binomial_fixed_variance(sim_df=tmp_df,\n",
    "                                                     mean_beta=mean_beta,\n",
    "                                                     var_beta=var_beta)\n",
    "        tmp_df[\"Pos\"] = pos\n",
    "        sim_list.append(tmp_df)\n",
    "\n",
    "    sim_df = pd.concat(sim_list, ignore_index=True)\n",
    "    return sim_df[\"het_level\"].mean()\n",
    "\n",
    "# Specify your desired mean heteroplasmy objective\n",
    "mean_het_objective = 0.7  # Replace this with your desired value\n",
    "\n",
    "# Set a tolerance level for the mean_het_objective\n",
    "tolerance = 0.1\n",
    "\n",
    "# Initialize lists to store optimization history\n",
    "mean_beta_history = []\n",
    "var_beta_history = []\n",
    "mean_het_history = []\n",
    "\n",
    "# Define the grid of mean_beta and var_beta values\n",
    "mean_beta_values = np.arange(0.00, 1.05, 0.05)\n",
    "mean_beta_values = [0.0, 0.1, 0.2, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]\n",
    "var_beta_values = [0.000001, 0.00001, 0.0001, 0.001, 0.01, 0.1]\n",
    "\n",
    "# Iterate over the grid\n",
    "for mean_beta in mean_beta_values:\n",
    "    for var_beta in var_beta_values:\n",
    "        current_mean_beta = mean_beta\n",
    "        current_var_beta = var_beta\n",
    "\n",
    "        # Initialize values\n",
    "        current_mean_het = calculate_mean_het(current_mean_beta, current_var_beta)\n",
    "\n",
    "        # Print statements for debugging\n",
    "        print(f\"Starting optimization for mean_beta={mean_beta}, var_beta={var_beta}\")\n",
    "\n",
    "        # Iterate until the mean_het_objective is reached\n",
    "        while np.abs(current_mean_het - mean_het_objective) > tolerance:\n",
    "            # Adjust mean_beta and var_beta\n",
    "            current_mean_beta += 0.05  # You can adjust the step size as needed\n",
    "            current_var_beta += 0.00001  # You can adjust the step size as needed\n",
    "\n",
    "            # Recalculate mean_het\n",
    "            current_mean_het = calculate_mean_het(current_mean_beta, current_var_beta)\n",
    "            print(f\"Current mean_het: {current_mean_het}\")\n",
    "\n",
    "        # Store the optimization results\n",
    "        mean_beta_history.append(current_mean_beta)\n",
    "        var_beta_history.append(current_var_beta)\n",
    "        mean_het_history.append(current_mean_het)\n",
    "\n",
    "# Plot the contour plot\n",
    "fig = plt.figure(figsize=(10, 8))\n",
    "ax = fig.add_subplot(111, projection='3d')\n",
    "\n",
    "# Create a grid of mean_beta and var_beta values\n",
    "mean_beta_grid, var_beta_grid = np.meshgrid(mean_beta_values, var_beta_values)\n",
    "\n",
    "# Calculate mean_het for each combination of mean_beta and var_beta\n",
    "mean_het_grid = np.array([[calculate_mean_het(mean_b, var_b) for mean_b in mean_beta_values] for var_b in var_beta_values])\n",
    "\n",
    "# Plot the contour plot\n",
    "ax.contour(mean_beta_grid, var_beta_grid, mean_het_grid, levels=20, cmap='viridis', label='Mean Heteroplasmy')\n",
    "\n",
    "# Plot the optimization path\n",
    "ax.scatter(mean_beta_history, var_beta_history, mean_het_history, color='red', marker='o', label='Optimization Path')\n",
    "\n",
    "ax.set_xlabel('Mean Beta')\n",
    "ax.set_ylabel('Var Beta')\n",
    "ax.set_zlabel('Mean Heteroplasmy')\n",
    "ax.set_title('Optimization Path for Mean Heteroplasmy')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make plots for sanity checking \n",
    "plt.figure(1, figsize=(18, 5))\n",
    "\n",
    "# # 1\n",
    "plt.subplot(131)\n",
    "plt.plot(sim_df[\"age\"], sim_df[\"coverage\"], '.k')\n",
    "plt.title('simulations: age vs. coverage', fontsize = 16, fontweight = \"bold\")\n",
    "plt.xlabel('age [y]', fontsize = 14, fontweight = \"bold\")\n",
    "plt.ylabel('coverage [reads]', fontsize = 14, fontweight = \"bold\")\n",
    "\n",
    "# # 2\n",
    "plt.subplot(132)\n",
    "plt.plot(sim_df[\"age\"], sim_df[\"het_level\"], '.k')\n",
    "plt.title('simulations: age vs. heteroplasmy', fontsize = 16, fontweight = \"bold\")\n",
    "plt.ylabel('heteroplasmy [%]', fontsize = 14, fontweight = \"bold\")\n",
    "plt.xlabel('age [y]', fontsize = 14, fontweight = \"bold\")\n",
    "plt.ylim(0, 1)\n",
    "\n",
    "# # 3\n",
    "plt.subplot(133)\n",
    "plt.plot(sim_df[\"coverage\"], sim_df[\"het_level\"], '.k')\n",
    "plt.ylabel('heteteroplasmy [%]', fontsize = 14, fontweight = \"bold\")\n",
    "plt.xlabel('coverage [reads]', fontsize = 14, fontweight = \"bold\")\n",
    "plt.title('simulations: coverage vs. heteroplasmy', fontsize = 16, fontweight = \"bold\")\n",
    "plt.ylim(0, 1)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## apply functions -- automated for production\n",
    "\n",
    "### simulate heteroplasmy values for selected params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params_df = pd.read_excel(\"/Users/simon.wengert/git/mtDNA_variants/metadata/params/2024_01_30_het_sim_params.xlsx\")\n",
    "params_df = params_df[params_df['run'] == True]\n",
    "params_df[\"n_donors\"] = params_df[\"n_donors\"].astype(int)\n",
    "params_df\n",
    "#params_df.drop(['change_param', 'step_param'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run simulations as specified in params_df\n",
    "save_path = \"/Users/simon.wengert/data/mtDNA_variants/metadata/cohort_files/gtex_v8/\"\n",
    "\n",
    "# Create an empty list to store simulated data frames\n",
    "simulated_dfs = []\n",
    "n_pos = 1000\n",
    "# Create a PDF to save all the plots\n",
    "date = \"2024_05_01_\"\n",
    "max_row = max(params_df[\"row_id\"].astype(int))\n",
    "for index, row in params_df.iterrows():\n",
    "    # tell me where we are:\n",
    "    print(f\"currently simulating row: {index} out of {max_row}\")\n",
    "    # Retrieve params\n",
    "    row_id = row[\"row_id\"]\n",
    "    change_param = row[\"change_param\"]\n",
    "    step_param = row[\"step_param\"]\n",
    "    n_donors = row['n_donors']\n",
    "    mean_age = row[\"mean_age\"]\n",
    "    sd_age = row[\"sd_age\"]\n",
    "    min_age_bound = row[\"min_age_bound\"]\n",
    "    max_age_bound = row[\"max_age_bound\"]\n",
    "    min_coverage = row[\"min_coverage\"]\n",
    "    slope = row[\"slope\"]\n",
    "    r_sqr = row[\"r_sqr\"]\n",
    "    mean_coverage = row[\"mean_coverage\"]\n",
    "    mean_beta = row[\"mean_beta\"]\n",
    "    var_beta = row[\"var_beta\"]\n",
    "    \n",
    "\n",
    "    #try:    \n",
    "    # init res list\n",
    "    sim_list = []\n",
    "    # loop through the number of positions you want to simulate\n",
    "    for pos in range(1, n_pos + 1):\n",
    "        # simulate coverage\n",
    "        tmp_df = sample_cohort_specs(n_donors=n_donors, \n",
    "                                     mean_age=mean_age,\n",
    "                                     sd_age=sd_age,\n",
    "                                     min_age_bound=min_age_bound,\n",
    "                                     max_age_bound=max_age_bound,\n",
    "                                     min_coverage=min_coverage,\n",
    "                                     slope=slope,\n",
    "                                     r_sqr=r_sqr,\n",
    "                                     mean_coverage=mean_coverage)\n",
    "        # simulate heteroplasmy inversely correlated with coverage using beta binomial\n",
    "        tmp_df = sample_beta_binomial_fixed_variance(sim_df=tmp_df, \n",
    "                                                     mean_beta=mean_beta, \n",
    "                                                     var_beta=var_beta)\n",
    "        # add Pos column \n",
    "        tmp_df[\"Pos\"] = pos\n",
    "        # append the result to the list\n",
    "        sim_list.append(tmp_df)\n",
    "\n",
    "    # concatenate the results into a single DataFrame\n",
    "    sim_df = pd.concat(sim_list, ignore_index=True)\n",
    "    \n",
    "    # Add param record to sim_df\n",
    "    sim_df[\"row_id\"] = row_id\n",
    "    sim_df[\"date\"] = date\n",
    "    sim_df[\"change_param\"] = change_param\n",
    "    sim_df[\"step_param\"] = step_param\n",
    "    sim_df[\"n_donors\"] = n_donors\n",
    "    sim_df[\"mean_age\"] = mean_age\n",
    "    sim_df[\"sd_age\"] = sd_age\n",
    "    sim_df[\"min_age_bound\"] = min_age_bound\n",
    "    sim_df[\"max_age_bound\"] = max_age_bound\n",
    "    sim_df[\"min_coverage\"] = min_coverage\n",
    "    sim_df[\"slope\"] = slope\n",
    "    sim_df[\"r_sqr\"] = r_sqr\n",
    "    sim_df[\"mean_coverage\"] = mean_coverage\n",
    "    sim_df[\"mean_beta\"] = mean_beta\n",
    "    sim_df[\"var_beta\"] = var_beta\n",
    "\n",
    "    #except Exception as e:\n",
    "    #   # Handle the exception (you can print an error message or log the error)\n",
    "    #    print(f\"Error in simulation for params: {row}\")\n",
    "    #    print(f\"Error message: {e}\")\n",
    "    #    continue  # Continue to the next iteration if an error occurs\n",
    "    \n",
    "    # Append the simulated DataFrame to the list\n",
    "    simulated_dfs.append(sim_df)\n",
    "\n",
    "# Merge all simulated data frames into one\n",
    "print(f\"concatenating simulations df --- \")   \n",
    "merged_sim_df = pd.concat(simulated_dfs)\n",
    "\n",
    "# Save the merged simulated DataFrame to disk\n",
    "print(f\"writing file to disk --- \") \n",
    "# merged_sim_df.to_csv(os.path.join(save_path, f\"{date}merged_simulated_data.tsv\"), sep=\"\\t\", index=False)#\n",
    "merged_sim_df.to_csv(os.path.join(save_path, f\"{date}merged_simulated_data_var_beta_large.tsv\"), sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# scratch ----------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load parameters from the text file into a DataFrame\n",
    "params_df = pd.read_csv(\"/Users/simon.wengert/git/mtDNA_variants/metadata/utils/het_sim_params_df_2.txt\", sep=\"\\t\")\n",
    "\n",
    "# Define absolute paths for saving files\n",
    "save_path = \"/Users/simon.wengert/data/mtDNA_variants/metadata/cohort_files/gtex_v8/\"\n",
    "figure_path = \"/Users/simon.wengert/git/mtDNA_variants/results/3_pheno_test/simulations/\"\n",
    "\n",
    "# Create an empty list to store simulated data frames\n",
    "simulated_dfs = []\n",
    "\n",
    "# Create a PDF to save all the plots\n",
    "date = \"2023_12_13_\"\n",
    "pdf_path = os.path.join(figure_path, f\"{date}heteroplasmy_simulation_plots_combined.pdf\")\n",
    "with PdfPages(pdf_path) as pdf:\n",
    "    for index, row in params_df.iterrows():\n",
    "    \n",
    "        # Retrieve params\n",
    "        row_id = row['row_id']\n",
    "        base_coverage = row['base_coverage']\n",
    "        slope = row['slope']\n",
    "        noise_factor = row['noise_factor']\n",
    "        mean_beta = row['mean_beta']\n",
    "        var_beta = row['var_beta']\n",
    "        \n",
    "        try:\n",
    "            # Simulate coverage\n",
    "            sim_df = sample_cohort_specs(base_coverage=base_coverage, slope=slope, noise_factor=noise_factor)\n",
    "            \n",
    "            # Simulate heteroplasmy inversely correlated with coverage using beta binomial\n",
    "            sim_df = sample_beta_binomial_fixed_variance(sim_df=sim_df, mean_beta=mean_beta, var_beta=var_beta)    \n",
    "\n",
    "            # Add param record to sim_df\n",
    "            sim_df[\"sim_param_row_id\"] = row_id\n",
    "            sim_df[\"sim_param_date\"] = date\n",
    "            sim_df['sim_param_base_coverage'] = base_coverage\n",
    "            sim_df['sim_param_slope'] = slope\n",
    "            sim_df['sim_param_noise_factor'] = noise_factor\n",
    "            sim_df['sim_param_mean_beta'] = mean_beta\n",
    "            sim_df['sim_param_var_beta'] = var_beta\n",
    "\n",
    "            # Joint title\n",
    "            joint_title = f\"Simulations: Base Coverage={base_coverage}, Slope={slope}, Noise Factor={noise_factor}, Mean Beta={mean_beta}, Var Beta={var_beta}\"\n",
    "\n",
    "            # Plot heteroplasmy simulations\n",
    "            plt.figure(figsize=(18, 5))\n",
    "\n",
    "            # # 1\n",
    "            plt.subplot(131)\n",
    "            plt.plot(sim_df[\"age\"], sim_df[\"coverage\"], '.k')\n",
    "            plt.title('Age vs. Coverage', fontsize=16, fontweight=\"bold\")\n",
    "            plt.xlabel('Age [y]', fontsize=14, fontweight=\"bold\")\n",
    "            plt.ylabel('Coverage [reads]', fontsize=14, fontweight=\"bold\")\n",
    "            plt.suptitle(joint_title, fontsize=18, fontweight=\"bold\")\n",
    "\n",
    "            # # 2\n",
    "            plt.subplot(132)\n",
    "            plt.plot(sim_df[\"age\"], sim_df[\"het_level\"], '.k')\n",
    "            plt.title('Age vs. Heteroplasmy', fontsize=16, fontweight=\"bold\")\n",
    "            plt.ylabel('Heteroplasmy [%]', fontsize=14, fontweight=\"bold\")\n",
    "            plt.xlabel('Age [y]', fontsize=14, fontweight=\"bold\")\n",
    "            plt.ylim(0, 1)\n",
    "\n",
    "            # # 3\n",
    "            plt.subplot(133)\n",
    "            plt.plot(sim_df[\"coverage\"], sim_df[\"het_level\"], '.k')\n",
    "            plt.ylabel('Heteroplasmy [%]', fontsize=14, fontweight=\"bold\")\n",
    "            plt.xlabel('Coverage [reads]', fontsize=14, fontweight=\"bold\")\n",
    "            plt.title('Coverage vs. Heteroplasmy', fontsize=16, fontweight=\"bold\")\n",
    "            plt.ylim(0, 1)\n",
    "            plt.tight_layout()\n",
    "\n",
    "            # Append the plot to the PDF\n",
    "            pdf.savefig(dpi=80)\n",
    "            plt.close()  # Close the current figure to avoid overlapping plots\n",
    "        \n",
    "        except Exception as e:\n",
    "            # Handle the exception (you can print an error message or log the error)\n",
    "            print(f\"Error in simulation for params: {row}\")\n",
    "            print(f\"Error message: {e}\")\n",
    "            continue  # Continue to the next iteration if an error occurs\n",
    "        \n",
    "        # Save the simulated DataFrame to disk\n",
    "        file_name = f\"{date}heteroplasmy_simulations_slope_{slope}_mean_beta_{mean_beta}_var_beta_{var_beta}.tsv\"\n",
    "        sim_df.to_csv(os.path.join(save_path, file_name), sep=\"\\t\", index=False)\n",
    "    \n",
    "        # Append the simulated DataFrame to the list\n",
    "        simulated_dfs.append(sim_df)\n",
    "\n",
    "# Merge all simulated data frames into one\n",
    "merged_sim_df = pd.concat(simulated_dfs)\n",
    "\n",
    "# Save the merged simulated DataFrame to disk\n",
    "merged_sim_df.to_csv(os.path.join(save_path, f\"{date}_merged_simulated_data.tsv\"), sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load parameters from the text file into a DataFrame\n",
    "params_df = pd.read_csv(\"/Users/simon.wengert/git/mtDNA_variants/metadata/utils/het_sim_params_df_2.txt\", sep=\"\\t\")\n",
    "\n",
    "# Define absolute paths for saving files\n",
    "save_path = \"/Users/simon.wengert/data/mtDNA_variants/metadata/cohort_files/gtex_v8/\"\n",
    "figure_path = \"/Users/simon.wengert/git/mtDNA_variants/results/3_pheno_test/simulations/\"\n",
    "\n",
    "# Create an empty list to store simulated data frames\n",
    "simulated_dfs = []\n",
    "\n",
    "# Loop through each row in the parameters DataFrame\n",
    "for index, row in params_df.iterrows():\n",
    "    \n",
    "    # Retrieve params\n",
    "    date = \"2023_12_13_\"\n",
    "    base_coverage = row['base_coverage']\n",
    "    slope = row['slope']\n",
    "    noise_factor = row['noise_factor']\n",
    "    mean_beta = row['mean_beta']\n",
    "    var_beta = row['var_beta']\n",
    "    \n",
    "    # Simulate coverage\n",
    "    sim_df = sample_cohort_specs(base_coverage=base_coverage, slope=slope, noise_factor=noise_factor)\n",
    "    \n",
    "    # Simulate heteroplasmy inversely correlated with coverage using beta binomial\n",
    "    sim_df = sample_beta_binomial_fixed_variance(sim_df=sim_df, mean_beta=mean_beta, var_beta=var_beta)    \n",
    "\n",
    "    # Add param record to sim_df\n",
    "    sim_df[\"sim_param_date\"] = date\n",
    "    sim_df['sim_param_base_coverage'] = base_coverage\n",
    "    sim_df['sim_param_slope'] = slope\n",
    "    sim_df['sim_param_noise_factor'] = noise_factor\n",
    "    sim_df['sim_param_mean_beta'] = mean_beta\n",
    "    sim_df['sim_param_var_beta'] = var_beta\n",
    "\n",
    "    # Joint title\n",
    "    joint_title = f\"Simulations: Base Coverage={base_coverage}, Slope={slope}, Noise Factor={noise_factor}, Mean Beta={mean_beta}, Var Beta={var_beta}\"\n",
    "\n",
    "    # Plot heteroplasmy simulations\n",
    "    plt.figure(figsize=(18, 5))\n",
    "\n",
    "    # # 1\n",
    "    plt.subplot(131)\n",
    "    plt.plot(sim_df[\"age\"], sim_df[\"coverage\"], '.k')\n",
    "    plt.title('Age vs. Coverage', fontsize=16, fontweight=\"bold\")\n",
    "    plt.xlabel('Age [y]', fontsize=14, fontweight=\"bold\")\n",
    "    plt.ylabel('Coverage [reads]', fontsize=14, fontweight=\"bold\")\n",
    "    plt.suptitle(joint_title, fontsize=18, fontweight=\"bold\")\n",
    "\n",
    "    # # 2\n",
    "    plt.subplot(132)\n",
    "    plt.plot(sim_df[\"age\"], sim_df[\"het_level\"], '.k')\n",
    "    plt.title('Age vs. Heteroplasmy', fontsize=16, fontweight=\"bold\")\n",
    "    plt.ylabel('Heteroplasmy [%]', fontsize=14, fontweight=\"bold\")\n",
    "    plt.xlabel('Age [y]', fontsize=14, fontweight=\"bold\")\n",
    "    plt.ylim(0, 1)\n",
    "\n",
    "    # # 3\n",
    "    plt.subplot(133)\n",
    "    plt.plot(sim_df[\"coverage\"], sim_df[\"het_level\"], '.k')\n",
    "    plt.ylabel('Heteroplasmy [%]', fontsize=14, fontweight=\"bold\")\n",
    "    plt.xlabel('Coverage [reads]', fontsize=14, fontweight=\"bold\")\n",
    "    plt.title('Coverage vs. Heteroplasmy', fontsize=16, fontweight=\"bold\")\n",
    "    plt.ylim(0, 1)\n",
    "    plt.tight_layout()\n",
    "\n",
    "    # Save the plot to a PDF file\n",
    "    plot_name = f\"{date}heteroplasmy_simulation_plots_slope_{slope}_mean_beta_{mean_beta}_var_beta_{var_beta}.pdf\"\n",
    "    plt.savefig(os.path.join(figure_path,plot_name))\n",
    "    plt.close()  # Close the current figure to avoid overlapping plots\n",
    "\n",
    "    # Save the simulated DataFrame to disk\n",
    "    file_name = f\"{date}heteroplasmy_simulations_slope_{slope}_mean_beta_{mean_beta}_var_beta_{var_beta}.tsv\"\n",
    "    sim_df.to_csv(os.path.join(save_path,file_name), sep=\"\\t\", index=False)\n",
    "    \n",
    "    # Append the simulated DataFrame to the list\n",
    "    simulated_dfs.append(sim_df)\n",
    "\n",
    "# Merge all simulated data frames into one\n",
    "merged_sim_df = pd.concat(simulated_dfs)\n",
    "\n",
    "# Save the merged simulated DataFrame to disk\n",
    "merged_sim_df.to_csv(os.path.join(save_path, \"{date}_merged_simulated_data.tsv\"), sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## session info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "session_info.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# scratch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2024_01_29: let's turn interactive code into a function\n",
    "def sample_cohort_specs(n_donors = 250, mean_age = 50, sd_age = 10, \n",
    "                        min_age_bound = 20, max_age_bound = 90, \n",
    "                        min_coverage = 10, max_coverage = 50000,\n",
    "                        slope_gradient = 0.1, noise_factor = 0.2, \n",
    "                        mean_coverage=1000):\n",
    "    \"\"\"\n",
    "    Generate synthetic cohort similar to GTEx data. The main point is to simulate\n",
    "    coverage values which are linearly dependent on donor age. The parameter \n",
    "    values are chosen such that the properties of the simulated cohort match\n",
    "    those of GTEx tissues best. These simulations are best thought of as \"per \n",
    "    tissue, per position\". \n",
    "\n",
    "    - donor ages are simulated using a truncated normal. \n",
    "    - coverage values are simulated using a linear model and Gaussian noise.\n",
    "    \n",
    "\n",
    "    n_donors:               number of donors to simulate \n",
    "    mean_age:               mean of the resulting age distribution\n",
    "    sd_age:                 standard deviation of the resulting age distribution\n",
    "    min_age_bound:          minimum of donor age distribution to simulate\n",
    "    max_age_bound:          minimum of donor age distribution to simulate\n",
    "    slope_gradient:         % of change in slope per year of age. Specified by user. \n",
    "    mean_coverage:          mean of coverage values to aim for when simulating.\n",
    "    min_coverage:           minimum coverage value found in real data.\n",
    "    noise_factor:           % of mean_coverage defining how much gaussian noise \n",
    "                            to add to age vs. coverage line.\n",
    "\n",
    "    \n",
    "    returns:   \"synthetic cohort\"  for tissue X and one position. Object is PD\n",
    "               containing columns age, mean_age, coverage, mean_coverage, coverage.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # simulate cohort ----------------------------------------------------------\n",
    "    # init empty data frame\n",
    "    sim_df = pd.DataFrame(np.zeros(n_donors),columns=[\"donor_id\"])\n",
    "    # generate donor_id \n",
    "    sim_df['donor_id'] = 'donor_' + (sim_df.index + 1).astype(str)\n",
    "\n",
    "\n",
    "    # simulate donor ages ------------------------------------------------------\n",
    "    # calculate the truncated normal bounds\n",
    "    a = (min_age_bound - mean_age) / sd_age\n",
    "    b = (max_age_bound - mean_age) / sd_age\n",
    "    # generate ages from a truncated normal distribution\n",
    "    ages = truncnorm.rvs(a, b, loc = mean_age, scale = sd_age, size = n_donors)\n",
    "    # round ages to integers\n",
    "    simulated_age = np.round(ages).astype(int)\n",
    "\n",
    "\n",
    "    # simulate coverage in linear dependence of donor age ----------------------\n",
    "    # the key question here is how to relate mean_coverage to our linear model\n",
    "    # slope coverage is defined as the [%] of change per year of life\n",
    "    slope_coverage = mean_coverage * slope_gradient \n",
    "    noise_coverage = mean_coverage * noise_factor\n",
    "    simulated_coverage = simulated_age * slope_coverage + np.random.normal(loc=0, scale=noise_coverage, size=len(sim_df))\n",
    "    # adjust for mean_coverage\n",
    "    simulated_coverage += mean_coverage - np.mean(simulated_coverage)\n",
    "    # set coverage values smaller than min_coverage to min_coverage \n",
    "    simulated_coverage = np.maximum(simulated_coverage, min_coverage)\n",
    "\n",
    "\n",
    "    # add return values to synthetic cohorts -----------------------------------\n",
    "    sim_df[\"age\"] = simulated_age\n",
    "    sim_df[\"mean_age\"] = np.mean(sim_df[\"age\"])\n",
    "    sim_df[\"coverage\"] = simulated_coverage\n",
    "    sim_df[\"mean_coverage\"] = np.mean(sim_df[\"coverage\"])\n",
    "    sim_df[\"coverage\"] = np.round(simulated_coverage).astype(int)\n",
    "\n",
    "\n",
    "    # return result ------------------------------------------------------------\n",
    "    return sim_df    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2024_01_29: let's turn interactive code into a function\n",
    "def sample_cohort_specs(n_donors = 250, mean_age = 50, sd_age = 10, \n",
    "                        min_age_bound = 20, max_age_bound = 90, \n",
    "                        intercept_coverage = 5000, slope_coverage = -400,\n",
    "                        sd_coverage = 3000, min_coverage = 10,\n",
    "                        max_coverage = 50000, mean_coverage=1000):#mean_coverage = 30000):\n",
    "    \"\"\"\n",
    "    Generate synthetic cohort similar to GTEx data. The main point is to simulate\n",
    "    coverage values which are linearly dependent on donor age. The parameter \n",
    "    values are chosen such that the properties of the simulated cohort match\n",
    "    those of GTEx tissues best. These simulations are best thought of as \"per \n",
    "    tissue, per position\". \n",
    "\n",
    "    - donor ages are simulated using a truncated normal. \n",
    "    - coverage values are simulated using a linear model and Gaussian noise.\n",
    "    \n",
    "\n",
    "    n_donors:               number of donors to simulate \n",
    "    mean_age:               mean of the resulting age distribution\n",
    "    sd_age:                 standard deviation of the resulting age distribution\n",
    "    min_age_bound:          minimum of donor age distribution to simulate\n",
    "    max_age_bound:          minimum of donor age distribution to simulate\n",
    "    intercept_coverage:     coverage intercept of the age ~ coverage simulation. \n",
    "                            Values are learned from lm() coefs in real data.\n",
    "    slope_coverage:         slope of the age ~ coverage simulation. lm() as above.\n",
    "    mean_coverage:          mean of coverage values to aim for when simulating.\n",
    "    sd_coverage:            std. deviation of coverage to aim for when simulating.\n",
    "    min_coverage:           minimum coverage value found in real data.\n",
    "\n",
    "    \n",
    "    returns:   \"synthetic cohort\"  for tissue X and one position. Object is PD\n",
    "               containing columns age, mean_age, coverage, mean_coverage, coverage.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # simulate cohort ----------------------------------------------------------\n",
    "    # init empty data frame\n",
    "    sim_df = pd.DataFrame(np.zeros(n_donors),columns=[\"donor_id\"])\n",
    "    # generate donor_id \n",
    "    sim_df['donor_id'] = 'donor_' + (sim_df.index + 1).astype(str)\n",
    "\n",
    "\n",
    "    # simulate donor ages ------------------------------------------------------\n",
    "    # calculate the truncated normal bounds\n",
    "    a = (min_age - mean_age) / sd_age\n",
    "    b = (max_age - mean_age) / sd_age\n",
    "    # generate ages from a truncated normal distribution\n",
    "    ages = truncnorm.rvs(a, b, loc = mean_age, scale = sd_age, size = n_donors)\n",
    "    # round ages to integers\n",
    "    simulated_age = np.round(ages).astype(int)\n",
    "\n",
    "\n",
    "    # simulate coverage in linear dependence of donor age ----------------------\n",
    "    # the key question here is how to relate mean_coverage to our linear model\n",
    "    # simulated_coverage = intercep_coverage + simulated_age * slope_coverage + np.random.normal(loc=0, scale=sd_coverage, size=len(sim_df))\n",
    "    simulated_coverage = simulated_age * slope_coverage + np.random.normal(loc=0, scale=sd_coverage, size=len(sim_df))\n",
    "    \n",
    "    # adjust for mean_coverage\n",
    "    simulated_coverage += mean_coverage - np.mean(simulated_coverage)\n",
    "    # set coverage values smaller than min_coverage to min_coverage \n",
    "    simulated_coverage = np.maximum(simulated_coverage, min_coverage)\n",
    "\n",
    "\n",
    "    # add return values to synthetic cohorts -----------------------------------\n",
    "    sim_df[\"age\"] = simulated_age\n",
    "    sim_df[\"mean_age\"] = np.mean(sim_df[\"age\"])\n",
    "    sim_df[\"coverage\"] = simulated_coverage\n",
    "    sim_df[\"mean_coverage\"] = np.mean(sim_df[\"coverage\"])\n",
    "    sim_df[\"coverage\"] = np.round(simulated_coverage).astype(int)\n",
    "\n",
    "\n",
    "    # return result ------------------------------------------------------------\n",
    "    return sim_df    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2024_01_29: --> archive this as previous: !!!! and do updates as discussed.\n",
    "\n",
    "# def sample_cohort_specs(n_donors = 2000, min_age = 20, max_age = 90, n_pos = 200, \n",
    "#                        base_coverage = 10, slope = 0.3 , noise_factor = 0.01):\n",
    "def sample_cohort_specs(n_donors = 2000, min_age = 20, max_age = 90, n_pos = 200, \n",
    "                        base_coverage = 10, intercept_factor=0.5, slope = 10, \n",
    "                        noise_factor = 15):\n",
    "    \"\"\"\n",
    "    Sample age for n_pos and n_donors from normal dist bounded by min_age and \n",
    "    max_age. Generate coverage count values from a linear relationship between \n",
    "    coverage and age.\n",
    "\n",
    "    Parameters:\n",
    "    n_donors (int): number of donors for simulation cohort. \n",
    "    min_age (int): lower age bound for sampling donor age.\n",
    "    max_age (int): upper age bound for sampling donor age.\n",
    "    n_pos (int): number of positions to initialize per donor and for which coverage\n",
    "    values will be sampled.\n",
    "    effect_cov_age (int): number of coverage read counts which will be added per \n",
    "    age-year in the linear model for sampling coverage. \n",
    "    noise_factor (int): numeric factor indicating the amount of random noise added\n",
    "    when simulation realtionship coverage with age.\n",
    "    \n",
    "    Returns:\n",
    "    pd.DataFrame: sim_df representing a mock cohort with donor_ids, evenly spaced\n",
    "    donor age, position_ids and coverage values which are positively correlated\n",
    "    with age.\n",
    "    \"\"\"\n",
    "    # init result df and aux params\n",
    "    sim_df = pd.DataFrame(np.zeros(n_donors),columns=[\"donor_id\"])\n",
    "\n",
    "    # simulate donor age as evenly spaced values \n",
    "    age = np.linspace(min_age, max_age, n_donors)\n",
    "    age = age.flatten()\n",
    "    age = np.round(age).astype(int)\n",
    "    sim_df[\"age\"] = age\n",
    "\n",
    "    # generate donor_id \n",
    "    sim_df['donor_id'] = 'donor_' + (sim_df.index + 1).astype(str)\n",
    "\n",
    "    # init positions\n",
    "    sim_df['pos_id'] = [list(range(1, 201)) for _ in range(len(sim_df))]\n",
    "    sim_df = sim_df.explode('pos_id', ignore_index=True)\n",
    "\n",
    "    # simulate coverage depending on age assuming positive linear relationship\n",
    "    # coverage = sim_df[\"age\"] * effect_cov_age + noise_factor * np.random.randn(len(sim_df)) - 900 # Paolo's way\n",
    "    # coverage = 1000 + sim_df[\"age\"] * effect_cov_age + noise_factor * np.random.randn(len(sim_df))\n",
    "    # coverage = sim_df[\"age\"] * effect_cov_age + noise_factor * np.random.randn(len(sim_df))  # let' use GTEx whole blood informed range of vals here  \n",
    "    # coverage = sim_df[\"age\"] + effect_cov_age + noise_factor * np.random.randn(len(sim_df))  # let' use GTEx whole blood informed range of vals here\n",
    "                \n",
    "    #coverage = np.random.normal(loc=base_coverage, scale=np.abs(sim_df['age'] * slope)) + noise_factor * np.random.randn(len(sim_df))  # let' use GTEx whole blood informed range of vals here  # SCALED COVERAGE!\n",
    "    # coverage = base_coverage * slope + np.random.normal(loc=0,scale = slope)\n",
    "    # coverage = 0 + sim_df['age'] *slope\n",
    "    # coverage = 0 + sim_df['age'] * slope + np.random.normal(loc=0, scale=slope * noise_factor, size=len(sim_df))\n",
    "    # coverage = base_coverage + sim_df['age'] * slope + np.random.normal(loc=0, scale=noise_factor, size=len(sim_df))\n",
    "    \n",
    "    # 2024_01_19: now flexibly adjust age coverage simulations proportional to specified values\n",
    "    adjusted_intercept=base_coverage*intercept_factor\n",
    "    adjusted_slope=(adjusted_intercept*slope)/base_coverage\n",
    "    adjusted_noise=noise_factor*adjusted_intercept\n",
    "    coverage = adjusted_intercept + sim_df['age'] * adjusted_slope + np.random.normal(loc=0, scale=adjusted_noise, size=len(sim_df))\n",
    "    \n",
    "    # Set negative values to zero\n",
    "    coverage = np.maximum(coverage, 0) # 10 potentially\n",
    "    # 2024_01_26: if < 0 do again or simulate more than you need and only take the ones above cutoff\n",
    "    \n",
    "    \n",
    "    # sim_df[\"coverage\"] = np.round(coverage).astype(int)\n",
    "    # sim_df[\"coverage_scaled\"] = coverage\n",
    "    sim_df[\"coverage\"] = coverage\n",
    "\n",
    "    #return results    \n",
    "    return sim_df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## apply functions -- manual\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set params manually for debugging before running automated pipeline above\n",
    "#     Pos tissue                pval mean_coverage base_coverage  slope mean_het   var_het      fdr\n",
    "# 1  1610 muscle_skeletal  0.00440            372.         559.   -3.72   0.139  0.0113    0.0396   \n",
    "#intercept_factor=0.5\n",
    "#base_coverage=4928\n",
    "# noise_factor=0.01\n",
    "#noise_factor = 0.05\n",
    "#slope=-36.151260\n",
    "mean_beta=0.027580\n",
    "#var_beta=0.011053\n",
    "var_beta=0.001337\n",
    "# var_beta=0.0019020\n",
    "# apply functions defined above\n",
    "# simulate coverage \n",
    "#sim_df = sample_cohort_specs(base_coverage = base_coverage, intercept_factor=intercept_factor, slope = slope, noise_factor = noise_factor)\n",
    "\n",
    "sim_list = []\n",
    "# number of simulations\n",
    "n_pos = 25\n",
    "# loop through the simulation process n_pos times\n",
    "for pos in range(1, n_pos + 1):\n",
    "    # simulate coverage\n",
    "    tmp_df = sample_cohort_specs()\n",
    "    # simulate heteroplasmy inversely correlated with coverage using beta binomial\n",
    "    tmp_df = sample_beta_binomial_fixed_variance(sim_df=tmp_df, mean_beta=mean_beta, var_beta=var_beta)\n",
    "    # add Pos column \n",
    "    tmp_df[\"Pos\"] = pos\n",
    "    # append the result to the list\n",
    "    sim_list.append(tmp_df)\n",
    "\n",
    "# concatenate the results into a single DataFrame\n",
    "sim_df = pd.concat(sim_list, ignore_index=True)\n",
    "\n",
    "# show contents sim_df \n",
    "sim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot heteroplasmy simultions\n",
    "plt.figure(1, figsize=(18, 5))\n",
    "\n",
    "# # 1\n",
    "plt.subplot(131)\n",
    "plt.plot(sim_df[\"age\"], sim_df[\"coverage\"], '.k')\n",
    "plt.title('simulations: age vs. coverage', fontsize = 16, fontweight = \"bold\")\n",
    "plt.xlabel('age [y]', fontsize = 14, fontweight = \"bold\")\n",
    "plt.ylabel('coverage [reads]', fontsize = 14, fontweight = \"bold\")\n",
    "\n",
    "# # 2\n",
    "plt.subplot(132)\n",
    "plt.plot(sim_df[\"age\"], sim_df[\"het_level\"], '.k')\n",
    "plt.title('simulations: age vs. heteroplasmy', fontsize = 16, fontweight = \"bold\")\n",
    "plt.ylabel('heteroplasmy [%]', fontsize = 14, fontweight = \"bold\")\n",
    "plt.xlabel('age [y]', fontsize = 14, fontweight = \"bold\")\n",
    "plt.ylim(0, 1)\n",
    "\n",
    "# # 3\n",
    "plt.subplot(133)\n",
    "plt.plot(sim_df[\"coverage\"], sim_df[\"het_level\"], '.k')\n",
    "plt.ylabel('heteteroplasmy [%]', fontsize = 14, fontweight = \"bold\")\n",
    "plt.xlabel('coverage [reads]', fontsize = 14, fontweight = \"bold\")\n",
    "plt.title('simulations: coverage vs. heteroplasmy', fontsize = 16, fontweight = \"bold\")\n",
    "plt.ylim(0, 1)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save file to disc for running the models on it in separate `R` session \n",
    "# (they are all implemented in `R`, that's why...)\n",
    "sim_df.to_csv(\"~/data/mtDNA_variants/metadata/cohort_files/gtex_v8/2023_12_13_test.tsv\", sep=\"\\t\")\n",
    "#sim_df.to_csv(\"~/data/mtDNA_variants/metadata/cohort_files/gtex_v8/2023_11_30_simulated_heteroplasmy_scaled_cov_slope_0_03_mean_het_0_5_scale_var_0_00000000000000000000000001.tsv\", sep=\"\\t\")\n",
    "#sim_df.to_csv(\"~/data/mtDNA_variants/metadata/cohort_files/gtex_v8/2023_11_30_simulated_heteroplasmy_age_on_cov_0_5_mean_het_0_5_scale_var_0_000005_lm_age_cov_and_overdisp.tsv\", sep=\"\\t\")\n",
    "#sim_df.to_csv(\"~/data/mtDNA_variants/metadata/cohort_files/gtex_v8/2023_11_28_simulated_heteroplasmy_mean_het_0_5_scale_var_0_15_no_cov_age_rel.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new figure and set up the first subplot\n",
    "plt.figure(figsize=(21, 5))\n",
    "\n",
    "# hist age\n",
    "plt.subplot(131)\n",
    "plt.hist(sim_df[\"age\"], bins=20, color='blue', alpha=0.7)\n",
    "plt.title('Age Distribution', fontsize=16, fontweight=\"bold\")\n",
    "plt.xlabel('Age', fontsize=14, fontweight=\"bold\")\n",
    "plt.ylabel('Frequency', fontsize=14, fontweight=\"bold\")\n",
    "\n",
    "# hist coverage\n",
    "plt.subplot(132)\n",
    "plt.hist(sim_df[\"coverage\"], bins=20, color='blue', alpha=0.7)\n",
    "plt.title('Coverage Distribution', fontsize=16, fontweight=\"bold\")\n",
    "plt.xlabel('Coverage', fontsize=14, fontweight=\"bold\")\n",
    "plt.ylabel('Frequency', fontsize=14, fontweight=\"bold\")\n",
    "\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SCRATCH ______________________________________________________________________"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sample_beta_binomial(sim_df, mean_het = 0.5, scale_var = 0.05):\n",
    "    \"\"\"\n",
    "    Sample heteroplasmic allele counts from a Beta-Binomial distribution. This \n",
    "    function is expected to be run on the outputs of sample_cohort_specs(). The \n",
    "    shape parameters of a Beta distribution are calculated from the desired mean\n",
    "    heteroplasmy (desired in terms of simulation goal) and the variance for the \n",
    "    Beta. This variance is scaled such that it is inversely correlated with \n",
    "    coverage (overdispersion). The heteroplasmy level is returned as the fraction\n",
    "    of simulated heteroplasmy counts over coverage.\n",
    "\n",
    "    Parameters:\n",
    "    sim_df (pd.DataFrame): simulated cohort data frame. Output of sample_cohort_specs().\n",
    "    mean_het (int): The \"true\" mean heteroplasmy level we assume when simulating \n",
    "    overdispersion.\n",
    "    scale_var (int): constant parameter which is used to scale the variance of the \n",
    "    Beta distribution inversely with coverage. \n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: sim_df is returned. Variance, alpha and beta shape parameters \n",
    "    for Beta distribution are returned. Also, p_beta, beta-estimated probability \n",
    "    for the Binomial draw to estimate heteroplasmy counts, and the heteroplasmy\n",
    "    allele counts and levels are returned as well. \n",
    "    \"\"\"\n",
    "    sim_df[\"scale_var\"] = scale_var\n",
    "    sim_df[\"var_beta\"] = (1/(sim_df[\"coverage\"]*sim_df[\"scale_var\"]))\n",
    "\n",
    "    # set mean to expected mean_het\n",
    "    sim_df[\"mean_het\"] = mean_het\n",
    "\n",
    "    # calculate alpha and beta from mean and variance\n",
    "    sim_df[\"common_factor\"] = sim_df[\"mean_het\"] * (1 - sim_df[\"mean_het\"]) / sim_df[\"var_beta\"] - 1\n",
    "    sim_df[\"alpha\"] = sim_df[\"mean_het\"] * sim_df[\"common_factor\"]\n",
    "    sim_df[\"beta\"] = (1 - sim_df[\"mean_het\"]) * sim_df[\"common_factor\"]\n",
    "\n",
    "    # sample from p_beta from Beta for using it in Bernoulli draw below\n",
    "    sim_df[\"p_beta\"] = np.random.beta(sim_df[\"alpha\"],sim_df[\"beta\"])\n",
    "\n",
    "    # sample from the Binomial distribution using the sampled Beta value\n",
    "    sim_df[\"het_counts\"] = np.random.binomial(sim_df[\"coverage\"], sim_df[\"p_beta\"])\n",
    "    sim_df[\"het_level\"] = sim_df[\"het_counts\"]/sim_df[\"coverage\"]\n",
    "        \n",
    "    return sim_df       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_beta=0.5\n",
    "var_beta=0.1\n",
    "expand_coverage_by=1000\n",
    "\n",
    "\n",
    "common_factor = sim_df[\"mean_het\"] * (1 - sim_df[\"mean_het\"]) / sim_df[\"var_beta\"] - 1\n",
    "alpha = sim_df[\"mean_het\"] * sim_df[\"common_factor\"]\n",
    "beta = (1 - sim_df[\"mean_het\"]) * sim_df[\"common_factor\"]\n",
    "\n",
    "# sample from p_beta from Beta for using it in Bernoulli draw below\n",
    "p_beta = np.random.beta(sim_df[\"alpha\"],sim_df[\"beta\"])\n",
    "\n",
    "# artifically expand_coverage from decimal values to integers since we \n",
    "# need integer counts for beta binomial regression \n",
    "coverage = sim_df[\"coverage_scaled\"] * expand_coverage_by\n",
    "coverage = np.round(coverage).astype(int)\n",
    "\n",
    "het_counts = np.random.binomial(coverage, sim_df[\"p_beta\"])\n",
    "het_level = sim_df[\"het_counts\"]/coverage\n",
    "\n",
    "het_level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# apply functions defined above\n",
    "\n",
    "# simulate mock cohort with linear age-coverage relationship\n",
    "# sim_df = sample_cohort_specs(effect_cov_age=0,noise_factor=120)\n",
    "# sim_df = sample_cohort_specs(effect_cov_age = 50, noise_factor = 100)\n",
    "# sim_df = sample_cohort_specs(effect_cov_age = 0.5, noise_factor = 2.5)\n",
    "# sim_df = sample_cohort_specs(base_coverage = 20, slope = 0.04 , noise_factor = 0.5)\n",
    "\n",
    "sim_df = sample_cohort_specs(base_coverage = 10, slope = 5, noise_factor = 15)\n",
    "\n",
    "# simulate heteroplasmy inversely correlated with coverage using beta binomial\n",
    "# sim_df = sample_beta_binomial(sim_df=sim_df, mean_het=0.2, scale_var=0.15)\n",
    "# sim_df = sample_beta_binomial(sim_df=sim_df, mean_het=0.5, scale_var=0.15)\n",
    "\n",
    "# simulate heteroplasmy not correlated with coverage using beta binomial with \n",
    "# constant variance (just for checking this)\n",
    "# sim_df = sample_beta_binomial_fixed_variance(sim_df=sim_df, mean_beta=0.5, var_beta=0.00000000000000000000000001)\n",
    "# sim_df = sample_beta_binomial_fixed_variance(sim_df=sim_df, mean_beta=0.5, var_beta=0.000005)\n",
    "\n",
    "\n",
    "sim_df = sample_beta_binomial_fixed_variance(sim_df=sim_df, mean_beta=0.1, var_beta=0.0000005)\n",
    "\n",
    "\n",
    "# show contents sim_df \n",
    "sim_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gene_tools",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
